File: inventory-service/src/main/java/com/catius/inventory/controller/InventoryController.java
package com.catius.inventory.controller;

import com.catius.inventory.controller.dto.request.InventoryRequest;
import com.catius.inventory.controller.dto.response.InventoryResponse;
import com.catius.inventory.service.InventoryService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/inventory")
@RequiredArgsConstructor
public class InventoryController {

    private final InventoryService inventoryService;

    @GetMapping("/{productId}")
    public InventoryResponse getStock(@PathVariable Long productId) {
        return inventoryService.findByProductId(productId);
    }

    @PostMapping("/reserve")
    public InventoryResponse reserve(@RequestBody InventoryRequest request) {
        return inventoryService.reserve(request);
    }

    @PostMapping("/release")
    public InventoryResponse release(@RequestBody InventoryRequest request) {
        return inventoryService.release(request);
    }
}

File: inventory-service/src/main/kotlin/com/catius/inventory/controller/InventoryController.kt
package com.catius.inventory.controller

import com.catius.inventory.controller.dto.*
import com.catius.inventory.domain.Inventory
import com.catius.inventory.domain.port.InventoryRepository
import com.catius.inventory.domain.usecase.FindInventoryUseCase
import com.catius.inventory.domain.usecase.ReleaseInventoryUseCase
import com.catius.inventory.domain.usecase.ReserveInventoryUseCase
import jakarta.validation.Valid
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.*
import java.time.Instant

@RestController
@RequestMapping("/api/v1/inventory")
class InventoryController(
    private val reserveInventoryUseCase: ReserveInventoryUseCase,
    private val releaseInventoryUseCase: ReleaseInventoryUseCase,
    private val findInventoryUseCase: FindInventoryUseCase,
    private val inventoryRepository: InventoryRepository,
) {

    @PostMapping("/reserve")
    fun reserve(@RequestBody @Valid request: ReserveRequest): ReserveResponse {
        val result = reserveInventoryUseCase.reserve(request.toCommand())
        return ReserveResponse.from(result)
    }

    @PostMapping("/release")
    fun release(@RequestBody @Valid request: ReleaseRequest) {
        releaseInventoryUseCase.release(request.toCommand())
    }

    @PostMapping("/increase")
    @Transactional
    fun increase(@RequestBody @Valid request: IncreaseRequest) {
        val inventory = inventoryRepository.findByProductIdForUpdate(request.productId!!)
            ?: Inventory(request.productId!!, 0, Instant.now())
        val updated = inventory.release(request.quantity!!, Instant.now())
        inventoryRepository.save(updated)
    }

    @GetMapping("/{productId}")
    fun find(@PathVariable productId: Long): InventoryResponse {
        val result = findInventoryUseCase.findByProductId(productId)
        return InventoryResponse.from(result)
    }
}
